lib/app/dispatcher.rb

```diff
@@ -0,0 +1,88 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'active_support'
+require 'concurrent-ruby'
+require 'connectors'
+require 'core'
+require 'utility'
+require 'app/config'
+
+module App
+  class Dispatcher
+    POLL_IDLING = (App::Config[:idle_timeout] || 60).to_i
+    TERMINATION_TIMEOUT = (App::Config[:termination_timeout] || 60).to_i
+
+    attr_reader :workers
+
+    def initialize
+      @workers = {}.with_indifferent_access
+      @pool = Concurrent::ThreadPoolExecutor.new(
+        min_threads: 5,
+        max_threads: 10,
+        max_queue: 100,
+        fallback_policy: :caller_runs
```

Comment:
Any particular reason for using `caller_runs`?

I would suggest `abort`, and the dispatcher catches it and logs it appropriately. We don't want the dispatcher thread to be blocked by a sync job, and it's acceptable a sync job is discarded if the workers are busy.

Reply:
Yeah I went back and forth on the strategy. In the end, I chose `caller_runs` as something that's like a poor man's backpressure solution, as in, by tying up the main thread it won't let other jobs that can't be processed be submitted. But I think for the initial implementation, we might be better off with the "fail early" strategy, and that would be aborting. I'll change it.